File: backend/src/interfaces/story.ts
import { model, Schema,Document,Types } from "mongoose"

export class Story extends Document{
    id: string;
    name: string;
    description?: string;
    thumbnailUrl: string;
    audioUrl: string;
    audioLength: number;
    publisherId: string;
    likes: string[];
    tags: string[];

    constructor({
        id,
        name,
        description,
        thumbnailUrl,
        audioUrl,
        audioLength,
        publisherId,
        likes = [],
        tags = []
    }: {
        id?: string;
        name: string;
        description?: string;
        thumbnailUrl: string;
        audioUrl: string;
        audioLength: number;
        publisherId: string;
        likes?: string[];
        tags?: string[];
    }) {
        super();
        this.id = id ?? new Types.ObjectId().toHexString();
        this.name = name;
        this.description = description;
        this.thumbnailUrl = thumbnailUrl;
        this.audioUrl = audioUrl;
        this.audioLength = audioLength;
        this.publisherId = publisherId;
        this.likes = likes;
        this.tags = tags;
    }

    copyWith({
        id,
        name,
        description,
        thumbnailUrl,
        audioUrl,
        audioLength,
        publisherId,
        likes,
        tags
    }: Partial<Story>): Story {
        return new Story({
            id: id ?? this.id,
            name: name ?? this.name,
            description: description ?? this.description,
            thumbnailUrl: thumbnailUrl ?? this.thumbnailUrl,
            audioUrl: audioUrl ?? this.audioUrl,
            audioLength: audioLength ?? this.audioLength,
            publisherId: publisherId ?? this.publisherId,
            likes: likes ?? this.likes,
            tags: tags ?? this.tags
        });
    }

    toMap(): Record<string, any> {
        return {
            id: this.id,
            name: this.name,
            description: this.description,
            thumbnailUrl: this.thumbnailUrl,
            audioUrl: this.audioUrl,
            audioLength: this.audioLength,
            publisherId: this.publisherId,
            likes: this.likes,
            tags: this.tags
        };
    }

    static fromMap(map: Record<string, any>): Story {
        return new Story({
            id: map.id,
            name: map.name,
            description: map.description,
            thumbnailUrl: map.thumbnailUrl,
            audioUrl: map.audioUrl,
            audioLength: map.audioLength,
            publisherId: map.publisherId,
            likes: map.likes,
            tags: map.tags
        });
    }

    toJson(): string {
        return JSON.stringify(this.toMap());
    }

    static fromJson(json: string): Story {
        return Story.fromMap(JSON.parse(json));
    }

    toString(): string {
        return `Story(id: ${this.id}, name: ${this.name}, description: ${this.description}, thumbnailUrl: ${this.thumbnailUrl}, audioUrl: ${this.audioUrl}, audioLength: ${this.audioLength}, publisherId: ${this.publisherId}, likes: ${this.likes}, tags: ${this.tags})`;
    }

    equals(other: Story): boolean {
        return this.id === other.id &&
            this.name === other.name &&
            this.description === other.description &&
            this.thumbnailUrl === other.thumbnailUrl &&
            this.audioUrl === other.audioUrl &&
            this.audioLength === other.audioLength &&
            this.publisherId === other.publisherId &&
            JSON.stringify(this.likes) === JSON.stringify(other.likes) &&
            JSON.stringify(this.tags) === JSON.stringify(other.tags);
    }

   
}

export const StorySchema = new Schema<Story>({
    id: { type: String, required: true, unique: true },
    name: { type: String, required: true },
    description: { type: String, required: false },
    thumbnailUrl: { type: String, required: true },
    audioUrl: { type: String, required: true },
    audioLength: { type: Number, required: true },
    publisherId: { type: String, required: true },
    likes: [{ type: String }],
    tags: [{ type: String }]
});

export const StoryModel = model<Story>('Story', StorySchema);

File: frontend/lib/models/story.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'dart:convert';

import 'package:collection/collection.dart';

class Story {
String id;
String name;
String description;
String thumbnailUrl;
String audioUrl;
String audioLength;
String publisherId;
String publisherName;
List<String> likes;
List<String> tags;
  Story({
    required this.id,
    required this.name,
    required this.description,
    required this.thumbnailUrl,
    required this.audioUrl,
    required this.audioLength,
    required this.publisherId,
    required this.publisherName,
    required this.likes,
    required this.tags,
  });


  Story copyWith({
    String? id,
    String? name,
    String? description,
    String? thumbnailUrl,
    String? audioUrl,
    String? audioLength,
    String? publisherId,
    String? publisherName,
    List<String>? likes,
    List<String>? tags,
  }) {
    return Story(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      thumbnailUrl: thumbnailUrl ?? this.thumbnailUrl,
      audioUrl: audioUrl ?? this.audioUrl,
      audioLength: audioLength ?? this.audioLength,
      publisherId: publisherId ?? this.publisherId,
      publisherName: publisherName ?? this.publisherName,
      likes: likes ?? this.likes,
      tags: tags ?? this.tags,
    );
  }

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'id': id,
      'name': name,
      'description': description,
      'thumbnailUrl': thumbnailUrl,
      'audioUrl': audioUrl,
      'audioLength': audioLength,
      'publisherId': publisherId,
      'publisherName': publisherName,
      'likes': likes,
      'tags': tags,
    };
  }

  factory Story.fromMap(Map<String, dynamic> map) {
    return Story(
      id: map['id'] as String,
      name: map['name'] as String,
      description: map['description'] as String,
      thumbnailUrl: map['thumbnailUrl'] as String,
      audioUrl: map['audioUrl'] as String,
      audioLength: map['audioLength'] as String,
      publisherId: map['publisherId'] as String,
      publisherName: map['publisherName'] as String,
      likes: List<String>.from((map['likes'] as List<String>),),
      tags: List<String>.from((map['tags'] as List<String>),),
    );
  }

  String toJson() => json.encode(toMap());

  factory Story.fromJson(String source) => Story.fromMap(json.decode(source) as Map<String, dynamic>);

  @override
  String toString() {
    return 'Story(id: $id, name: $name, description: $description, thumbnailUrl: $thumbnailUrl, audioUrl: $audioUrl, audioLength: $audioLength, publisherId: $publisherId, publisherName: $publisherName, likes: $likes, tags: $tags)';
  }

  @override
  bool operator ==(covariant Story other) {
    if (identical(this, other)) return true;
    final listEquals = const DeepCollectionEquality().equals;
  
    return 
      other.id == id &&
      other.name == name &&
      other.description == description &&
      other.thumbnailUrl == thumbnailUrl &&
      other.audioUrl == audioUrl &&
      other.audioLength == audioLength &&
      other.publisherId == publisherId &&
      other.publisherName == publisherName &&
      listEquals(other.likes, likes) &&
      listEquals(other.tags, tags);
  }

  @override
  int get hashCode {
    return id.hashCode ^
      name.hashCode ^
      description.hashCode ^
      thumbnailUrl.hashCode ^
      audioUrl.hashCode ^
      audioLength.hashCode ^
      publisherId.hashCode ^
      publisherName.hashCode ^
      likes.hashCode ^
      tags.hashCode;
  }
}
